File: Semester4/Advanced Programming/Lab2_Hw/src/main/java/repository/EntitiesRepos/TicketDBRepository.java
package repository.EntitiesRepos;

import domain.entities.Ticket;
import domain.exceptions.IllegalArgumentException;
import domain.exceptions.ValidatorException;
import domain.validator.Validator;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import repository.CRUDRepository;
import repository.DataBaseUtils.DBUtils;

import java.sql.*;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class TicketDBRepository implements CRUDRepository<Integer, Ticket> {
    private DBUtils dbUtils;
    private Validator<Ticket> ticketValidator;
    private static final Logger logger = LogManager.getLogger("Ticket Logger");

    /**
     * Constructor for Ticket DataBase Repository
     * @param props - properties file
     * @param ticketValidator - validator for Ticket class
     */
    public TicketDBRepository(Properties props, Validator<Ticket> ticketValidator){
        logger.info("Initializing TicketDBRepo.... {}", props);
        this.dbUtils = new DBUtils(props);
        this.ticketValidator = ticketValidator;
    }


    /**
     * Function to find an Ticket from the database
     * @param integer - id of the Ticket
     * @return - the entity with the given id if it exists, null otherwise
     * @throws IllegalArgumentException - if the id is null
     */
    @Override
    public Ticket findOne(Integer integer) throws IllegalArgumentException {
        logger.traceEntry("Entered FindOne : IdTicket {}", integer);
        if(integer == null){
            throw  new IllegalArgumentException("Ticket id can't be null");
        }
        Connection con = dbUtils.getConnection();
        try(PreparedStatement prSt = con.prepareStatement("select * from Ticket where ticketId= ?")){
            prSt.setInt(1, integer);
            try(ResultSet res = prSt.executeQuery()){
                if(res.next()){
                    Integer id = res.getInt("ticketId");
                    Integer concId = res.getInt("concertId");
                    String name = res.getString("clientName");
                    LocalDate date = res.getDate("buyDate").toLocalDate();
                    Integer seats = res.getInt("numOfSeats");
                    logger.traceExit("Exited findOne: Ticket");
                    return new Ticket(id, concId, name, date, seats);
                }
            }
        } catch (SQLException e) {
            logger.error("Error in Ticket find ... {}", e);
            System.out.println("Error in TicketDB " + e);
        }

        logger.traceExit("There is no Ticket with the id... {}", integer);
        return null;
    }

    /**
     * Function to save ane entity into the database
     * @param entity - entity to be saved
     * @return - null if the entity was saved, the given entity otherwise
     * @throws IllegalArgumentException - if entity is null
     * @throws ValidatorException - if the entity is not well formed
     */
    @Override
    public Ticket save(Ticket entity) throws IllegalArgumentException, ValidatorException {
        logger.traceEntry("Entered Save Ticket.... {}", entity);
        if(entity == null){
            throw new IllegalArgumentException("Ticket can't be null");
        }
        ticketValidator.validate(entity);

        if(this.findOne(entity.getId()) != null){
            logger.info("Ticket with given id already in repo... {}", entity.getId());
            return entity;
        }

        Connection con = dbUtils.getConnection();
        try(PreparedStatement prSt = con.prepareStatement("insert into Ticket values(?, ?, ?, ?, ?)")){
            prSt.setInt(1, entity.getId());
            prSt.setInt(2, entity.getConcertId());
            prSt.setString(3, entity.getClientName());
            prSt.setDate(4, Date.valueOf(entity.getBuyDate()));
            prSt.setInt(5, entity.getNumOfSeats());
            int ret = prSt.executeUpdate();
            if(ret == 0){
                return entity;
            }
        } catch (SQLException e) {
            logger.error("Error in Ticket save.... {}", e);
            System.out.println("Error in TicketDB "  + e);
            return entity;
        }

        logger.traceExit("Ticket successfully saved");
        return null;
    }

    /**
     * Function to delete an entity from the repository
     * @param integer - id of the entity to be deleted
     * @return - the deleted entity if it was deleted, null otherwise
     * @throws IllegalArgumentException -if the id is null
     */
    @Override
    public Ticket delete(Integer integer) throws IllegalArgumentException {
        logger.traceEntry("Entered Ticket delete... {}", integer);
        if(integer == null){
            throw new IllegalArgumentException("Ticket id can't be null");
        }

        Ticket Ticket = findOne(integer);
        if(Ticket == null){
            logger.info("There is no Ticket with id... {}", integer);
            return null;
        }

        Connection con = dbUtils.getConnection();
        try(PreparedStatement prSt = con.prepareStatement("delete from Ticket where ticketId = ?")){
            prSt.setInt(1, integer);
            prSt.executeUpdate();
            logger.traceExit("Exited Ticket delete.... {}", Ticket);
            return Ticket;
        } catch (SQLException e) {
            logger.error("Error in Ticket delete.... {}", e);
            e.printStackTrace();
        }

        logger.info("Couldn't delete the given Ticket ....{}", Ticket);
        return null;
    }

    /**
     * Function to update an entity from the repository
     * @param entity - the entity to be updated
     * @return - null if the entity was updated, the given entity otherwise
     * @throws IllegalArgumentException - if entity is null
     * @throws ValidatorException - if the given entity is not well formed
     */
    @Override
    public Ticket update(Ticket entity) throws IllegalArgumentException, ValidatorException {
        logger.traceEntry("Entered Update Ticket.... {}", entity);
        if(entity == null){
            throw new IllegalArgumentException("The updated Ticket can't be null");
        }
        ticketValidator.validate(entity);

        if(this.findOne(entity.getId()) == null){
            logger.info("Ticket with given id not in repo... {}", entity.getId());
            return entity;
        }

        Connection con = dbUtils.getConnection();
        try(PreparedStatement prSt = con.prepareStatement("update Ticket set concertId=?, clientName=?, buyDate=?, numOfSeats = ?" +
                " where TicketId = ?")){
            prSt.setInt(5, entity.getId());
            prSt.setInt(1, entity.getConcertId());
            prSt.setString(2, entity.getClientName());
            prSt.setDate(3, Date.valueOf(entity.getBuyDate()));
            prSt.setInt(4, entity.getNumOfSeats());

            int ret = prSt.executeUpdate();
            if(ret == 0){
                return entity;
            }
        } catch (SQLException e) {
            logger.error("Error in Ticket update.... {}", e);
            System.out.println("Error in TicketDB "  + e);
        }

        logger.traceExit("Ticket successfully updated");
        return null;
    }

    /**
     * Function to get all entities from the repository
     * @return - iterable containing all entities from the repository
     */
    @Override
    public Iterable<Ticket> findAll() {
        logger.entry("Entered in findAll Tickets");
        List<Ticket> TicketList = new ArrayList<>();
        Connection con = dbUtils.getConnection();
        try(PreparedStatement prSt = con.prepareStatement("select * from Ticket")){
            try(ResultSet res = prSt.executeQuery()){
                while(res.next()){
                    Integer id = res.getInt("ticketId");
                    Integer concId = res.getInt("concertId");
                    String name;
                    name = res.getString("clientName");
                    LocalDate date = res.getDate("buyDate").toLocalDate();
                    Integer seats = res.getInt("numOfSeats");

                    TicketList.add(new Ticket(id, concId, name, date, seats));
                }
            }
        } catch (SQLException e) {
            logger.error("Error in findAll Tickets... {}", e);
            System.out.println("Error in db Tickets... " + e);
        }

        logger.traceExit("Exited findAll Tickets");
        return TicketList;
    }

    /**
     * Function to retrieve the number of entities from the database
     * @return - number of entity in the database
     */
    @Override
    public Integer size() {
        logger.traceEntry("Entered Ticket size");
        Connection con = dbUtils.getConnection();

        try(PreparedStatement prSt = con.prepareStatement("select count(*) as size from Ticket")){
            try(ResultSet res = prSt.executeQuery()){
                if(res.next()){
                    logger.traceExit("Exited Ticket size");

                    return res.getInt("size");
                }
            }
        } catch (SQLException e) {
            logger.error("Error in Ticket size  ...{}", e);
            System.out.println("Error in Ticket DB " + e);
        }

        logger.info("No entries in db");
        return 0;
    }
}

File: Semester4/Advanced Programming/ThriftProj/C#Thrift/Lab2_Festival/Server/repository/entitiesRepos/TicketDBRepository.cs
using Lab2_Festival.domain.entities;
using Lab2_Festival.domain.validators;
using Lab2_Festival.repository.dbUtils;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Lab2_Festival.repository.entitiesRepos
{
    public class TicketDBRepository : ICRUDRepository<int, Ticket>
    {
        private readonly IValidator<Ticket> _validator;
        private readonly IDictionary<string, string> _props;
        private IDictionary<string, string> props;
        private TicketValidator ticketValidator;

        public TicketDBRepository(IDictionary<string, string> props, IValidator<Ticket> validator) 
        {
            this._props = props;
            this._validator = validator;
        }

        public TicketDBRepository(IDictionary<string, string> props, TicketValidator ticketValidator)
        {
            this.props = props;
            this.ticketValidator = ticketValidator;
        }

        public Ticket Delete(int id)
        {
            throw new NotImplementedException();
        }

        public IEnumerable<Ticket> FindAll()
        {
            IList<Ticket> tickets = new List<Ticket>();

            var conn = DBUtil.GetConnection(_props);

            using(var comm = conn.CreateCommand())
            {
                comm.CommandText = "select * from ticket";

                using(var dataR = comm.ExecuteReader())
                {
                    while (dataR.Read())
                    {
                        int ticketId = dataR.GetInt32(0);
                        int concertId = dataR.GetInt32(1);
                        string clientName = dataR.GetString(2);
                        DateTime buyDate = dataR.GetDateTime(3);
                        int numOfSeats = dataR.GetInt32(4);
                        
                        tickets.Add(new Ticket(ticketId, concertId, clientName, buyDate.ToString("MM/dd/yyyy"), numOfSeats));
                    }
                }
            }
            return tickets;
        }

        public Ticket FindOne(int id)
        {
            var con = DBUtil.GetConnection(_props);
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select ticketId, concertId, clientName, buyDate, numOfSeats from ticket where ticketId=@id";
                var paramId = comm.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = id;
                comm.Parameters.Add(paramId);

                using (var datR = comm.ExecuteReader())
                {
                    if (datR.Read())
                    {
                        int ticketId = datR.GetInt32(0);
                        int concertId = datR.GetInt32(1);
                        string clientName = datR.GetString(2);
                        var date = datR.GetDateTime(3);
                        var numOfSeats = datR.GetInt32(4);
                        //TODO change this
                        
                        return new Ticket(ticketId, concertId, clientName, date.ToString("MM/dd/yyyy"), numOfSeats);
                       
                    }
                }
            }
            return null;
        }

        public Ticket Save(Ticket e)
        {
            var conn = DBUtil.GetConnection(_props);
     
            var artist = FindOne(e.Id);
            if (artist != null)
            {
                return e;
            }

            using (var comm = conn.CreateCommand())
            {
                comm.CommandText = "insert into ticket values(@id, @concertId, @clientName, @buyDate, @numOfSeats)";

                var tickId = comm.CreateParameter();
                tickId.ParameterName = "@id";
                tickId.Value = e.Id;
                comm.Parameters.Add(tickId);


                var concId = comm.CreateParameter();
                concId.ParameterName = "@concertId";
                concId.Value = e.ConcertId;
                comm.Parameters.Add(concId);

                var clName = comm.CreateParameter();
                clName.ParameterName = "@clientName";
                clName.Value = e.ClientName;
                comm.Parameters.Add(clName);

                var buyDate = comm.CreateParameter();
                buyDate.ParameterName = "@buyDate";
                buyDate.Value = Convert.ToDateTime(e.BuyDate);
                comm.Parameters.Add(buyDate);

                var numSeats = comm.CreateParameter();
                numSeats.ParameterName = "numOfSeats";
                numSeats.Value = e.NumOfSeats;
                comm.Parameters.Add(numSeats);


                var res = comm.ExecuteNonQuery();

                return null;

            }
        }

        public int Size()
        {
            var nrTickets = 0;
            var conn = DBUtil.GetConnection(_props);
            using (var comm = conn.CreateCommand())
            {
                comm.CommandText = "select * from ticket";

                using (var dataR = comm.ExecuteReader())
                {
                    while (dataR.Read())
                    {
                        nrTickets++;
                    }
                }
            }

            return nrTickets;
        }

        public Ticket Update(Ticket e)
        {
            throw new NotImplementedException();
        }
    }
}
